### PR TITLE
Avoid circular type definition

### LINKED ISSUE #57: Remove circular type definition
The circular type definition for `PortAudioStream` and `PortAudioSource`/`Sink` is clever but less inferable and not strictly necessary:

Instead, we could have a nested structure:

- `Portal` (or some other name) would would contain a `PortAudioDevice`, buffer, and number of channels
- `PortAudioStream` would contain input and output portals and the pointer to the stream
- `PortAudioSource`/`Sink` would be trivial wrappers over `PortAudioStream` to show what context to use the stream in

### PR DESCRIPTION
**Pull request by bramtayl**
See #57 for motivation
Closes #57

### CODE REVIEW

File: src/PortAudio.jl [hunk 1]
```diff
@@ -290,16 +303,16 @@ function PortAudioStream(fn::Function, args...; kwargs...)
...
 function close(stream::PortAudioStream)
-    if stream.stream != C_NULL
-        Pa_StopStream(stream.stream)
-        Pa_CloseStream(stream.stream)
-        stream.stream = C_NULL
+    pointer = stream.pointer_ref[]
+    if stream.pointer_ref[] != C_NULL
+        Pa_StopStream(stream.pointer_ref[])
+        Pa_CloseStream(stream.pointer_ref[])
+        stream.pointer_ref[] = C_NULL
```
Comment (bramtayl) [inline src/PortAudio.jl:310]: It seems like we should just be able to set the pointer itself to 0? That would remove the need for a ref altogether. I followed the original code here.

Comment (bramtayl) [reply src/PortAudio.jl:310]: Oh, nevermind, I think I get why that wouldn't work. Pointers are immutable; only what they point to can change?

**Comment by rob-luke** [comment]
I'll be back in the lab next week and can give his a listen to then unless ssfr reviews in the meantime.

**Comment by bramtayl** [comment]
Been a week so merging

**Pull request merged**